application.cpp

```diff
@@ -203,10 +217,16 @@ void application::start_sighup_handler( std::shared_ptr<boost::asio::signal_set>
 #endif
 }
 
+std::unique_ptr<application> application::app_instance; // static
+
 application& application::instance() {
-   static application _app;
-   return _app;
+   if (__builtin_expect((app_instance && !app_instance->should_reset), 1))
+      return *app_instance;
+   app_instance.reset(nullptr); // delete old application first
+   app_instance.reset(new application);
+   return *app_instance;
```

Comment:
I'm not sure this approach will work. The `io_service` has an accessor `get_io_service()` which returns a reference to the internal io context. 

I think instead we need a `restart()` or `reexec()` method that re-registers the plugins, initializes them, and then calls `exec()`. This approach should remove the need to destroy/re-create the application object and avoid having to add this `if` statement. Also, the sig handling can likely be left as-is.

We will need to remove the `io_serv.reset()` from `exec()`, but that shouldn't need to be there anyway.

Reply:
> I'm not sure this approach will work. The io_service has an accessor get_io_service() which returns a reference to the internal io context.

Yes, it does, but why do you think it wouldn't work? The application is recreated only after the io_service has been destroyed.

Reply:
But there are parts of the code that still have and could potentially use the reference.

Reply:
After the application::exec() terminated and we executed `application::shutdown()` and `io_serv.reset();` If that is the case , anyone using a reference would crash anyways I believe and I think it would be a bug.

Reply:
Yes, it is a bug if we allow application to be re-used, but that is the point of this exercise.

Reply:
Talked with Kevin in person. We'll have to make sure that if the `io_service` created in `application::application()` is used outside of plugins (for example in `gelf_appender`), we reinitialize these when recreating a new application. A good spot for doing this may be in the `chain_plugin` initialization.

Plugins are created when the `application` is created, and destroyed when the `application` is destroyed. They are initialized (by calling `plugin_initialize`) either:
* when `application::initialize<plugin, ...>` is called. plugins withing the template parameter list are initialized. 
* or within `application::initialize()` if --plugin is specified as a command line argument

Reply:
Is there a reason we can't state that the `io_context` returned via `get_io_service()` is invalid after all plugins shutdown? gelf_appender is a pretty interesting example -- it would violate the rule today but it seems rather trivial to refactor it to not violate the rule?

Reply:
> refactor it to not violate the rule

You mean refactor it as a plugin, so that it gets managed by appbase and destroyed before the io_service is deleted. I think that would indeed be the correct way of addressing the issue.

Reply:
The easiest thing to do would be for gelf_appender to have its own io_context. The downside is, just doing that alone, means would need to remove the `async_send_to()` fallbacks in it. But if the socket buffer size was bumped up.. maybe these asyncs really aren't such a hard requirement since there will be very little back pressure on a UDP socket anyways.

More complex would be for gelf_appender to run its own thread with its own io_context, so it could continue to support async fallback. I get that it feels overkill (and I kinda think it is).

Not sure, are there other examples of this other then gelf_appender? If it's really just gelf_appender maybe it's worth cleaning it up so don't need to do something kludgy in chain_plugin.